pkg/audit/manager.go

```diff
@@ -1061,19 +1064,33 @@ func emitEvent(constraint *unstructured.Unstructured,
 		logging.ResourceNamespace:    rnamespace,
 		logging.ResourceName:         rname,
 	}
+
 	reason := "AuditViolation"
-	ref := getViolationRef(gkNamespace, resourceGroupVersionKind.Kind, rname, rnamespace, constraint.GetKind(), constraint.GetName(), constraint.GetNamespace())
+	enamespace := gkNamespace
+	if *auditEventsInvolvedNamespace && len(rnamespace) > 0 {
+		enamespace = rnamespace
+	}
+
+	ref := getViolationRef(enamespace, resourceGroupVersionKind.Kind, rname, rrv, ruid)
 
-	eventRecorder.AnnotatedEventf(ref, annotations, corev1.EventTypeWarning, reason, "Timestamp: %s, Resource Namespace: %s, Constraint: %s, Message: %s", timestamp, rnamespace, constraint.GetName(), message)
+	if *auditEventsInvolvedNamespace || len(rnamespace) == 0 {
+		eventRecorder.AnnotatedEventf(ref, annotations, corev1.EventTypeWarning, reason, "Constraint: %s, Message: %s", constraint.GetName(), message)
+	} else {
+		eventRecorder.AnnotatedEventf(ref, annotations, corev1.EventTypeWarning, reason, "Resource Namespace: %s, Constraint: %s, Message: %s", rnamespace, constraint.GetName(), message)
+	}
 }
 
-func getViolationRef(gkNamespace, rkind, rname, rnamespace, ckind, cname, cnamespace string) *corev1.ObjectReference {
-	return &corev1.ObjectReference{
+func getViolationRef(enamespace, rkind, rname, rrv string, ruid types.UID) *corev1.ObjectReference {
+	ref := &corev1.ObjectReference{
 		Kind:      rkind,
 		Name:      rname,
-		UID:       types.UID(rkind + "/" + rnamespace + "/" + rname + "/" + ckind + "/" + cnamespace + "/" + cname),
-		Namespace: gkNamespace,
+		Namespace: enamespace,
+	}
+	if len(ruid) > 0 && len(rrv) > 0 {
```

Comment:
Is there a case where you only want to set one of them if only one of them is valid?

Reply:
tbh I'm not sure, I thought about it initially but I couldn't think why a resource version would not exist if the object has been assigned a UID.